p2p/src/authenticated/discovery/actors/listener.rs

```diff
@@ -145,17 +238,187 @@ impl<E: Spawner + Clock + ReasonablyRealtime + Network + Rng + CryptoRng + Metri
+
+            // Drop the connection if the IP exceeds its rate limit
+            if self.ip_rate_limiter.check_key(&address.ip()).is_err() {
+                self.handshakes_ip_rate_limited.inc();
+                debug!(ip = ?address.ip(), "ip exceeded handshake rate limit");
+                self.prioritized_ips.write().await.remove(&address.ip());
+                continue;
+            }
+
+            if self.subnet_rate_limiter.check_key(&subnet).is_err() {
+                self.handshakes_subnet_rate_limited.inc();
+                debug!(ip = ?address.ip(), subnet = ?subnet, "subnet exceeded handshake rate limit");
+                self.prioritized_ips.write().await.remove(&address.ip());
+                continue;
+            }
+
+            // Attempt to reserve a slot for the handshake
+            let in_flight = loop {
+                let result = self.in_flight_handshakes.fetch_update(
+                    Ordering::AcqRel,
+                    Ordering::Relaxed,
+                    |current| (current < self.max_concurrent_handshakes).then_some(current + 1),
+                );
+                match result {
+                    Ok(_) => break InFlightGuard::new(self.in_flight_handshakes.clone()),
+                    Err(_) if prioritized_ip => {
+                        self.context.sleep(Duration::from_millis(1)).await;
+                    }
+                    Err(_) => {
+                        self.handshakes_rate_limited.inc();
+                        debug!(?address, "maximum concurrent handshakes reached");
+                        self.prioritized_ips.write().await.remove(&address.ip());
+                        continue 'accept;
+                    }
+                }
+            };
+
             // Spawn a new handshaker to upgrade connection
             self.context.with_label("handshaker").spawn({
                 let stream_cfg = self.stream_cfg.clone();
                 let tracker = tracker.clone();
                 let supervisor = supervisor.clone();
+                let peer_rate_limiter = self.peer_rate_limiter.clone();
+                let prioritized_ips = self.prioritized_ips.clone();
+                let handshakes_peer_rate_limited = self.handshakes_peer_rate_limited.clone();
+                let in_flight = in_flight;
                 move |context| {
                     Self::handshake(
-                        context, address, stream_cfg, sink, stream, tracker, supervisor,
+                        context,
+                        address,
+                        stream_cfg,
+                        sink,
+                        stream,
+                        tracker,
+                        supervisor,
+                        peer_rate_limiter,
+                        prioritized_ips,
+                        handshakes_peer_rate_limited,
+                        in_flight,
                     )
                 }
             });
         }
     }
 }
+
+struct InFlightGuard {
```

Comment:
nit: move this to utils (as a "best-effort" semaphore = concurrency guard) that returns a marker if acquired but doesn't block on it.